### ISSUE TITLE
[FX] `GraphFunction` class (standalone function support)

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [pytorch/pytorch#56712]
## 🚀 Feature
It would be nice if it were possible to take advantage of the power of FX to process and generate free functions as well as `torch.nn.Module`s (`fx.GraphModule`), likely by adding some kind of `fx.GraphFunction` class. This would mirror the distinction between `torch.jit.ScriptModule` and `torch.jit.ScriptFunction`; ideally `torch.jit.script(my_graph_function)` would yield a `ScriptFunction` when `my_graph_function` is an `fx.GraphFunction`.

## Motivation

Sometimes code generated by FX doesn't need to access module buffers, submodules, etc. but is a simple function of the inputs.
## Pitch

An `fx.GraphFunction` class as well as a way to get the `.python_code` of a graph without `self` as a standalone function. (The first will use the second, I'd assume.)

## Alternatives

Currently, to do this, we have to use a hack like:
```python
code = graph.python_code('')
code = code.replace("def forward(self, ", "def main(")
tmp_globals = {}
fx.graph_module.exec_with_source(code, tmp_globals)
main = tmp_globals["main"]
```

**Comment by Chillee** [comment]
Could you explain a bit about why you need this? I understand the intuitive appeal of passing around functions instead of modules, but I'm not totally sure what the advantage of this is.

**Comment by Linux-cpp-lisp** [comment]
Sure! It's related to issues I'm having with mixed serialization of TorchScript and Python modules. (See [this post](https://github.com/pytorch/pytorch/issues/18106#issuecomment-825166656), for example.)

We have a library that generates a lot of low level code for our networks using `fx`, and we want this code to always be compiled to TorchScript for efficiency.  At the same time, we *don't* want to compile the whole model to TorchScript all of the time — there are methods, attributes, etc. that can be important for training and analysis that are not supported in TorchScript. Obviously, we want to be able to save our models. When the internal compiled code exists as a `ScriptFunction`, I'm able to use `__getstate__`/`__setstate__`  to manage serialization, save the source code string but not the `ScriptFunction`, and make it work. When it is a `ScriptModule`, on the other hand, preventing `Mixed serialization not supported` errors requires much uglier hacks.

Obviously the above is not a very common usecase (except for users of our library), but it made me realize that it would be nice to have an officially supported way to get a standalone function ('s code) for an `fx.Graph` without hacky string manipulation and using the innards of FX.

**Comment by Chillee** [comment]
There is another issue with the idea of transforming pure functions - there are some cases where we need to store state onto the tensor (i.e. `tensor_constant_`). The issue is that the original function can have arbitrary logic (call into numpy, read global variables, etc.), but the FX module we call needs to be "closed" - it can't access the same global state and stuff like that.

There is a somewhat related issue that I've been thinking about, which is that you can't currently call a FX function from within another one.

For example

```
def f(x): ...
f = fx.symbolic_trace(f)

def g(x):
   return f(x)*2
```
This currently fails because `f` becomes a module.
